BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt

```diff
@@ -0,0 +1,343 @@
+import android.provider.Settings
+import android.view.Menu
+import android.view.MenuItem
+import android.view.View
+import androidx.appcompat.app.AppCompatActivity
+import androidx.core.app.ActivityCompat
+import androidx.core.widget.TextViewCompat
+import com.google.android.gms.auth.api.signin.GoogleSignIn
+import com.google.android.gms.fit.samples.common.logger.Log
+import com.google.android.gms.fit.samples.common.logger.LogView
+import com.google.android.gms.fit.samples.common.logger.LogWrapper
+import com.google.android.gms.fit.samples.common.logger.MessageOnlyLogFilter
+import com.google.android.gms.fitness.Fitness
+import com.google.android.gms.fitness.FitnessOptions
+import com.google.android.gms.fitness.data.DataSource
+import com.google.android.gms.fitness.data.DataType
+import com.google.android.gms.fitness.request.DataSourcesRequest
+import com.google.android.gms.fitness.request.OnDataPointListener
+import com.google.android.gms.fitness.request.SensorRequest
+import com.google.android.material.snackbar.Snackbar
+import java.util.concurrent.TimeUnit
+
+const val TAG = "BasicSensorsApi"
+
+/**
+ * This sample demonstrates how to use the Sensors API of the Google Fit platform to find available
+ * data sources and to register/unregister listeners to those sources. It also demonstrates how to
+ * authenticate a user with Google Play Services.
+ */
+class MainActivity : AppCompatActivity() {
+    private val fitnessOptions = FitnessOptions.builder().addDataType(DataType.TYPE_LOCATION_SAMPLE).build()
+
+    // Two maps are used in conjunction with the OAuth flow. This is in order to identify what
+    // function to execute once the flow has successfully completed.
+    private val signInActionMap = mapOf<() -> Any, Int>(
+            ::findFitnessDataSources to 1
+    )
+    private val signInResponseMap = signInActionMap.entries.associate { (k, v) -> v to k }
+
+    // [START dataPointListener_variable_reference]
+    // Need to hold a reference to this listener, as it's passed into the "unregister"
+    // method in order to stop all sensors from sending data to this listener.
+    private var dataPointListener: OnDataPointListener? = null
+
+    // [END dataPointListener_variable_reference]
+
+    // [START auth_oncreate_setup]
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        // Put application specific code here.
+        setContentView(R.layout.activity_main)
+        // This method sets up our custom logger, which will print all log messages to the device
+        // screen, as well as to adb logcat.
+        initializeLogging()
+        // When permissions are revoked the app is restarted so onCreate is sufficient to check for
+        // permissions core to the Activity's functionality.
+        checkPermissionsAndRun(::findFitnessDataSources)
+    }
+
+    private fun checkPermissionsAndRun(signInAction: () -> Any) {
+        if (hasRuntimePermissions()) {
```

Comment:
Refactor "hasRuntimePermissions()" to "permissionsApproved()" as it's much easier to understand.